base/boot.jl

```diff
@@ -270,13 +270,46 @@ end
 macro inline()   Expr(:meta, :inline)   end
 macro noinline() Expr(:meta, :noinline) end
 
+import .Intrinsics: ult_int
+struct Summarized
+    desc::String
+    size
+    type
+    function Summarized(@nospecialize a)
+        # try to get a summary of `a` here now rather than capturing it for later inspection,
+        # in order to allow compiler analyses or optimization passes to assume the invariant
+        # that this `BoundsError` doesn't escape `a` when it is a certain primitive object
+        # that they can reason about
+        if isa(a, Array) || isa(a, Tuple)
+            if isdefined(Main, :Base)
+                Base = Main.Base
+                # these `invoke_in_world`s here essentially prohibits users from changing
+                # the semantics of `BoundsError` by overriding `summary(a)` implementations
+                # (and potentially allowing `a` to be escaped to somewhere)
+                desc = Base.invoke_in_world(Base._BASE_DEFINED_AGE, Base.summary, a)::String
```

Comment:
I am a bit worried about that. When I joked offline that we "supported" this, @JeffBezanson was "We do not!".

I think a better approach is to define a `Core.throw_bounds_error` that is defined in the inference world-age. 
and have `jl_throw_bounds_error` invoke in that world-age.

Reply:
I don't think we really need to (or should) call `summary` here; saving the size and type should be enough.

To handle calling `size`, we could define constructors here only for Array and Tuple, and add a constructor in Base for other types so it can call Base.size. (I hope to have a better general solution for this one day, but for now this is how it is.)

Reply:
`size` -> `axes`